File: repro/p-b-habitat-percolation-ecology-fss/simulate_percolation_fss.py
#!/usr/bin/env python3
"""Crosscheck repro: 2D site percolation finite-size scaling of p_c(L).

Lattice:        L × L square
Neighborhood:   4-neighbor (von Neumann)
Boundary:       periodic (torus)
Spanning rule:  wrapping in either direction (horizontal or vertical)
Algorithm:      Newman–Ziff union-find (one occupation sequence per sample)
Estimator:      p_c(L) = mean occupation fraction at first wrapping
                (named: mean first either-wrap). SE = sample SE of the mean.

Theory:  p_c(∞) = 0.59274621,  ν = 4/3
Model:   p_c(L) = p_c(∞) + c L^{-1/ν}

Fit L ∈ {32, 64, 128, 256}. L=16 is printed as a diagnostic (corrections-to-
scaling) and is not used in the exponent fit.

This protocol measures the finite-size *shift* of an effective threshold, not
a generic spanning-probability demo. Open-boundary Π_TB = 0.5 is the wrong
estimator here: Cardy's result says that crossing probability on a square
tends to 1/2 at p_c(∞), so the leading L^{-1/ν} amplitude is ~0 and a
noisy four-point log|Δp| fit cannot recover ν.

Pass / fail / inconclusive (stdout RESULT token; exit code is always 0):
  CONFIRMED     weighted Fit A recovers ν within 15% of 4/3, with SE(ν)
                inside that tolerance, significant c, and R² ≥ 0.85
  INCONCLUSIVE  underpowered (SE too large, sign flips, or R² too low).
                Not a claim that percolation theory is wrong.
  FALSIFIED     powered, and ν is more than 15% from 4/3 with 4/3 outside 2σ

The in-browser JS runner is a smoke test and must not emit CONFIRMED.

CI greps stdout for RESULT: CONFIRMED on a powered run.
"""
from __future__ import annotations

import argparse
import math
import random
import sys
import time
from typing import Dict, List, Sequence, Tuple

PC_INF = 0.59274621
NU_THEORY = 4 / 3
NU_TOLERANCE = 0.15
R2_MIN = 0.85
FIT_SIZES = [32, 64, 128, 256]
DIAGNOSTIC_SIZES = [16]
N_SAMPLES = 400
SEED = 42

# Legacy (pre-fix) demo / DFS-bisection numbers, kept for the power note.
LEGACY_DEMO_TRIALS_PER_P = 120
LEGACY_PYTHON_TRIALS_PER_P = 350
LEGACY_SIZES = [16, 32, 64, 128]
# 2026-06-22 "CONFIRMED" snapshot stored in tests — noisy, L=64 ≈ p_c(∞).
LEGACY_REFERENCE_PCS = [0.59080, 0.59059, 0.59268, 0.59179]


def first_wrap_either(L: int, rng: random.Random) -> float:
    """Occupation fraction at first either-direction wrap on an L×L torus."""
    n = L * L
    parent = list(range(n))
    dx = [0] * n
    dy = [0] * n
    rank = [0] * n
    occupied = [False] * n

    def find(x: int) -> int:
        if parent[x] != x:
            orig = parent[x]
            root = find(orig)
            dx[x] += dx[orig]
            dy[x] += dy[orig]
            parent[x] = root
            return root
        return x

    wrap_h = False
    wrap_v = False
    order = list(range(n))
    rng.shuffle(order)
    for k, s in enumerate(order, start=1):
        occupied[s] = True
        r, c = divmod(s, L)
        bonds = (
            ((r - 1) % L, c, -1, 0),
            ((r + 1) % L, c, 1, 0),
            (r, (c - 1) % L, 0, -1),
            (r, (c + 1) % L, 0, 1),
        )
        for nr, nc, by, bx in bonds:
            t = nr * L + nc
            if not occupied[t]:
                continue
            rs, rt = find(s), find(t)
            if rs == rt:
                wx = dx[s] + bx - dx[t]
                wy = dy[s] + by - dy[t]
                if wx != 0:
                    wrap_h = True
                if wy != 0:
                    wrap_v = True
            elif rank[rs] < rank[rt]:
                parent[rs] = rt
                dx[rs] = dx[t] - bx - dx[s]
                dy[rs] = dy[t] - by - dy[s]
            else:
                parent[rt] = rs
                dx[rt] = dx[s] + bx - dx[t]
                dy[rt] = dy[s] + by - dy[t]
                if rank[rs] == rank[rt]:
                    rank[rs] += 1
        if wrap_h or wrap_v:
            return k / n
    return 1.0


def sample_pc(L: int, n_samples: int, rng: random.Random) -> Tuple[List[float], float, float, float]:
    """Return (samples, mean, SE of mean, stdev)."""
    ps = [first_wrap_either(L, rng) for _ in range(n_samples)]
    mean = sum(ps) / n_samples
    var = sum((p - mean) ** 2 for p in ps) / (n_samples - 1) if n_samples > 1 else 0.0
    sigma = math.sqrt(var)
    se = sigma / math.sqrt(n_samples) if n_samples else float("inf")
    return ps, mean, se, sigma


def _se_nu_from_chi2(
    chi2_of: Sequence[Tuple[float, float]], nu_hat: float, chi2_min: float
) -> float:
    """1σ width from the χ² = χ²min + 1 contour (1 parameter)."""
    target = chi2_min + 1.0
    left = chi2_of[0][0]
    right = chi2_of[-1][0]
    prev_nu, prev_chi = chi2_of[0]
    found_left = False
    found_right = False
    for nu, chi2 in chi2_of[1:]:
        if (not found_left) and prev_nu <= nu_hat and prev_chi >= target > chi2:
            denom = prev_chi - chi2
            frac = (prev_chi - target) / denom if denom else 0.0
            left = prev_nu + frac * (nu - prev_nu)
            found_left = True
        if (not found_right) and prev_nu >= nu_hat and prev_chi < target <= chi2:
            denom = chi2 - prev_chi
            frac = (target - prev_chi) / denom if denom else 0.0
            right = prev_nu + frac * (nu - prev_nu)
            found_right = True
        prev_nu, prev_chi = nu, chi2
    if not found_left:
        left = chi2_of[0][0]
    if not found_right:
        right = chi2_of[-1][0]
    return max(nu_hat - left, right - nu_hat)


def wls_through_origin(
    x: Sequence[float], y: Sequence[float], se: Sequence[float]
) -> Tuple[float, float, float, float]:
    """Weighted least squares y = c x (no intercept). Returns c, se_c, chi2, R²."""
    w = [1.0 / (s * s) if s > 0 else 0.0 for s in se]
    den = sum(wi * xi * xi for wi, xi in zip(w, x))
    if den <= 0:
        return 0.0, float("inf"), float("inf"), 0.0
    c = sum(wi * xi * yi for wi, xi, yi in zip(w, x, y)) / den
    se_c = math.sqrt(1.0 / den)
    resid = [yi - c * xi for xi, yi in zip(x, y)]
    chi2 = sum(wi * ri * ri for wi, ri in zip(w, resid))
    y_mean = sum(y) / len(y)
    ss_tot = sum((yi - y_mean) ** 2 for yi in y)
    ss_res = sum(ri * ri for ri in resid)
    r2 = 1.0 - ss_res / ss_tot if ss_tot else 0.0
    return c, se_c, chi2, r2


def fit_a(
    sizes: Sequence[int], pcs: Sequence[float], ses: Sequence[float]
) -> Dict[str, float]:
    """Fix p_c(∞), fit c and ν in p_c = p_c(∞) + c L^{-1/ν}."""
    y = [pc - PC_INF for pc in pcs]
    nu_grid = [0.50 + i * 0.002 for i in range(int((2.50 - 0.50) / 0.002) + 1)]
    best = None
    chi2_of: List[Tuple[float, float]] = []
    for nu in nu_grid:
        x = [L ** (-1.0 / nu) for L in sizes]
        c, se_c, chi2, r2 = wls_through_origin(x, y, ses)
        chi2_of.append((nu, chi2))
        if best is None or chi2 < best[0]:
            best = (chi2, nu, c, se_c, r2)
    assert best is not None
    chi2_min, nu_hat, c_hat, se_c, r2 = best
    se_nu = _se_nu_from_chi2(chi2_of, nu_hat, chi2_min)
    return {
        "nu": nu_hat,
        "se_nu": se_nu,
        "c": c_hat,
        "se_c": se_c,
        "chi2": chi2_min,
        "r2": r2,
        "rel_err": abs(nu_hat - NU_THEORY) / NU_THEORY,
    }


def fit_b(
    sizes: Sequence[int], pcs: Sequence[float], ses: Sequence[float]
) -> Dict[str, float]:
    """Fix ν = 4/3, fit c in p_c = p_c(∞) + c L^{-3/4}."""
    y = [pc - PC_INF for pc in pcs]
    x = [L ** (-1.0 / NU_THEORY) for L in sizes]
    c, se_c, chi2, r2 = wls_through_origin(x, y, ses)
    return {"c": c, "se_c": se_c, "chi2": chi2, "r2": r2, "nu": NU_THEORY}


def fit_width_nu(sizes: Sequence[int], sigmas: Sequence[float]) -> Tuple[float, float]:
    """Sanity: σ(p_span) ~ L^{-1/ν} via unweighted log-log. Returns (ν, R²)."""
    xs = [math.log(L) for L in sizes]
    ys = [math.log(s) for s in sigmas]
    n = len(xs)
    x_mean = sum(xs) / n
    y_mean = sum(ys) / n
    den = sum((x - x_mean) ** 2 for x in xs)
    if den <= 0:
        return float("inf"), 0.0
    slope = sum((x - x_mean) * (y - y_mean) for x, y in zip(xs, ys)) / den
    nu = -1.0 / slope if slope else float("inf")
    ss_res = sum((y - (y_mean + slope * (x - x_mean))) ** 2 for x, y in zip(xs, ys))
    ss_tot = sum((y - y_mean) ** 2 for y in ys)
    r2 = 1.0 - ss_res / ss_tot if ss_tot else 0.0
    return nu, r2


def same_sign_deltas(pcs: Sequence[float]) -> bool:
    deltas = [pc - PC_INF for pc in pcs]
    if any(d == 0 for d in deltas):
        return False
    signs = [d > 0 for d in deltas]
    return all(signs) or not any(signs)


def approaches_pc(sizes: Sequence[int], pcs: Sequence[float], ses: Sequence[float]) -> bool:
    """|p_c(L) - p_c(∞)| shrinks with L, allowing 2σ noise."""
    abs_d = [abs(pc - PC_INF) for pc in pcs]
    for i in range(len(sizes) - 1):
        slack = 2.0 * (ses[i] + ses[i + 1])
        if abs_d[i + 1] > abs_d[i] + slack:
            return False
    return True


def classify(fit_a_res: Dict[str, float], pcs: Sequence[float], ses: Sequence[float]) -> str:
    powered = fit_a_res["se_nu"] <= NU_TOLERANCE * NU_THEORY
    r2_ok = fit_a_res["r2"] >= R2_MIN
    slope_sig = abs(fit_a_res["c"]) > 2.0 * fit_a_res["se_c"]
    within = fit_a_res["rel_err"] <= NU_TOLERANCE
    sign_ok = same_sign_deltas(pcs)
    mono = approaches_pc(FIT_SIZES, pcs, ses) if len(pcs) == len(FIT_SIZES) else sign_ok

    if not sign_ok or not powered or not slope_sig or not r2_ok:
        return "INCONCLUSIVE"
    if within and mono:
        return "CONFIRMED"
    if (not within) and abs(fit_a_res["nu"] - NU_THEORY) > 2.0 * fit_a_res["se_nu"]:
        return "FALSIFIED"
    return "INCONCLUSIVE"


def samples_for_tolerance(n_samples: int, se_nu: float) -> int:
    target = NU_TOLERANCE * NU_THEORY
    if se_nu <= 0 or se_nu == float("inf"):
        return n_samples
    return max(n_samples, int(math.ceil(n_samples * (se_nu / target) ** 2)))


def legacy_power_note() -> str:
    """Why the old 120-trial open-Π=0.5 demo was INCONCLUSIVE."""
    width_128 = 128 ** (-1.0 / NU_THEORY)
    se_pi = math.sqrt(0.25 / LEGACY_DEMO_TRIALS_PER_P)
    # dΠ/dp ~ L^{1/ν}; Cardy Π(p_c)=1/2 so the *shift* amplitude is ~0.
    se_pc_rough = se_pi * (128 ** (1.0 / NU_THEORY)) / 8.0
    return (
        "Legacy demo (open TB spanning, noisy independent-seed bisection, "
        f"{LEGACY_DEMO_TRIALS_PER_P} trials/p, L∈{LEGACY_SIZES}):\n"
        f"  transition width at L=128 is ~L^{{-3/4}} ≈ {width_128:.4f}\n"
        f"  SE(Π) at Π=0.5 with {LEGACY_DEMO_TRIALS_PER_P} Bernoulli trials ≈ {se_pi:.3f}\n"
        f"  rough SE(p_c) ≳ {se_pc_rough:.4f}, comparable to (or larger than) the\n"
        "  expected shift because Cardy puts Π_TB(p_c(∞)) → 1/2, so c ≈ 0.\n"
        "  log|Δp| vs log L is unstable when a point (e.g. L=64) has Δp ≈ 0.\n"
        f"  The 2026-06-22 Python snapshot at {LEGACY_PYTHON_TRIALS_PER_P} trials/p was\n"
        f"  p_c = {LEGACY_REFERENCE_PCS} — non-monotonic, L=64 on top of p_c(∞).\n"
        "  INCONCLUSIVE there is a power/estimator bug, not a disproof of ν=4/3."
    )


def parse_args(argv: Sequence[str] | None = None) -> argparse.Namespace:
    p = argparse.ArgumentParser(description="2D site percolation FSS of p_c(L)")
    p.add_argument("--samples", type=int, default=N_SAMPLES)
    p.add_argument("--seed", type=int, default=SEED)
    return p.parse_args(argv)


def main(argv: Sequence[str] | None = None) -> int:
    args = parse_args(argv)
    n_samples = args.samples
    seed = args.seed
    all_sizes = DIAGNOSTIC_SIZES + FIT_SIZES
    t0 = time.perf_counter()

    print("Crosscheck: p-b-habitat-percolation-ecology-fss")
    print("Lattice: L×L square | neighborhood: 4-neighbor | BC: periodic")
    print("Estimator: mean occupation at first either-direction wrapping")
    print("Algorithm: Newman–Ziff union-find")
    print(f"Theory: p_c(inf)={PC_INF}, nu={NU_THEORY:.4f}")
    print(f"Params: L in {all_sizes}, N_SAMPLES={n_samples}, seed={seed}")
    print(f"Fit uses L={FIT_SIZES}; L={DIAGNOSTIC_SIZES} diagnostic only")
    print()
    print(legacy_power_note())
    print()

    means: Dict[int, float] = {}
    ses: Dict[int, float] = {}
    sigmas: Dict[int, float] = {}
    rng = random.Random(seed)
    for L in all_sizes:
        _ps, mean, se, sigma = sample_pc(L, n_samples, rng)
        means[L] = mean
        ses[L] = se
        sigmas[L] = sigma
        delta = mean - PC_INF
        x = L ** (-1.0 / NU_THEORY)
        flag = "fit" if L in FIT_SIZES else "diag"
        print(
            f"  L={L:4d}  p_c_hat={mean:.5f} ± {se:.5f}  "
            f"delta={delta:+.5f}  L^(-3/4)={x:.5f}  sigma={sigma:.5f}  [{flag}]"
        )

    fit_sizes = FIT_SIZES
    fit_pcs = [means[L] for L in fit_sizes]
    fit_ses = [ses[L] for L in fit_sizes]
    fit_sig = [sigmas[L] for L in fit_sizes]

    a = fit_a(fit_sizes, fit_pcs, fit_ses)
    b = fit_b(fit_sizes, fit_pcs, fit_ses)
    nu_w, r2_w = fit_width_nu(fit_sizes, fit_sig)
    result = classify(a, fit_pcs, fit_ses)
    need = samples_for_tolerance(n_samples, a["se_nu"])
    wall = time.perf_counter() - t0
    n_trials = n_samples * len(all_sizes)

    print()
    print("Fit A  (fix p_c(inf), fit c and nu; weighted by SE):")
    print(
        f"  nu = {a['nu']:.4f} ± {a['se_nu']:.4f}   "
        f"c = {a['c']:+.4f} ± {a['se_c']:.4f}"
    )
    print(
        f"  R² = {a['r2']:.4f}   chi² = {a['chi2']:.3f}   "
        f"rel err vs 4/3 = {100 * a['rel_err']:.1f}%  (tolerance {100 * NU_TOLERANCE:.0f}%)"
    )
    print("Fit B  (fix nu=4/3, fit c):")
    print(
        f"  c = {b['c']:+.4f} ± {b['se_c']:.4f}   "
        f"R² = {b['r2']:.4f}   chi² = {b['chi2']:.3f}"
    )
    print(
        f"Width sanity  sigma ~ L^(-1/nu):  nu = {nu_w:.4f}  (R² = {r2_w:.4f})"
    )
    print()
    print(
        f"Power: SE(nu)={a['se_nu']:.4f}; "
        f"target SE ≤ {NU_TOLERANCE * NU_THEORY:.4f} "
        f"needs ~{need} samples/L (current {n_samples})."
    )
    if result == "INCONCLUSIVE":
        print("INCONCLUSIVE means this run is underpowered or the fit is unstable.")
        print("It is not a claim that 2D percolation has the wrong nu.")
    elif result == "FALSIFIED":
        print("FALSIFIED means this estimator+budget disagrees with nu=4/3.")
        print("That is a protocol failure, not a disproof of percolation theory.")
    print(f"Wall-clock {wall:.1f}s  independent sequences={n_trials}")
    print()
    extra = {
        "CONFIRMED": "",
        "INCONCLUSIVE": " (increase --samples; demo JS cannot confirm nu)",
        "FALSIFIED": " (powered disagreement with nu=4/3 for this estimator)",
    }[result]
    print(f"RESULT: {result}{extra}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: repro/p-b-habitat-percolation-ecology-fss/simulate_percolation_fss.js
/**
 * In-browser Crosscheck for 2D site percolation finite-size scaling.
 *
 * Same estimator, lattices, sample count, and 15% gate as
 * simulate_percolation_fss.py. The RESULT line is the weighted fit.
 * It is not preset.
 *
 * Estimator id: mean-first-either-wrap
 * p_c(L) = mean occupation fraction at first wrapping in either direction.
 */
(function () {
  "use strict";

  const PC_INF = 0.59274621;
  const NU_THEORY = 4 / 3;
  const NU_TOLERANCE = 0.15;
  const R2_MIN = 0.85;
  const FIT_SIZES = [32, 64, 128, 256];
  const N_SAMPLES = 400;
  const SEED = 42;
  const ESTIMATOR_ID = "mean-first-either-wrap";
  const ESTIMATOR_SENTENCE =
    "2D site percolation, periodic 4-neighbor square lattice, Newman–Ziff; p_c(L) is the mean occupation fraction at first wrapping in either direction.";

  function mulberry32(seed) {
    return function () {
      seed |= 0;
      seed = (seed + 0x6d2b79f5) | 0;
      let t = Math.imul(seed ^ (seed >>> 15), 1 | seed);
      t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
      return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
    };
  }

  function yieldToBrowser() {
    return new Promise((resolve) => setTimeout(resolve, 0));
  }

  function shuffle(arr, rng) {
    for (let i = arr.length - 1; i > 0; i--) {
      const j = Math.floor(rng() * (i + 1));
      const tmp = arr[i];
      arr[i] = arr[j];
      arr[j] = tmp;
    }
  }

  function firstWrapEither(L, rng) {
    const n = L * L;
    const parent = new Int32Array(n);
    const dx = new Int32Array(n);
    const dy = new Int32Array(n);
    const rank = new Int32Array(n);
    const occupied = new Uint8Array(n);
    for (let i = 0; i < n; i++) parent[i] = i;

    function find(x) {
      if (parent[x] !== x) {
        const orig = parent[x];
        const root = find(orig);
        dx[x] += dx[orig];
        dy[x] += dy[orig];
        parent[x] = root;
        return root;
      }
      return x;
    }

    const order = new Int32Array(n);
    for (let i = 0; i < n; i++) order[i] = i;
    shuffle(order, rng);

    let wrapH = false;
    let wrapV = false;
    for (let k = 0; k < n; k++) {
      const s = order[k];
      occupied[s] = 1;
      const r = (s / L) | 0;
      const c = s - r * L;
      const bonds = [
        [(r - 1 + L) % L, c, -1, 0],
        [(r + 1) % L, c, 1, 0],
        [r, (c - 1 + L) % L, 0, -1],
        [r, (c + 1) % L, 0, 1],
      ];
      for (const [nr, nc, by, bx] of bonds) {
        const t = nr * L + nc;
        if (!occupied[t]) continue;
        const rs = find(s);
        const rt = find(t);
        if (rs === rt) {
          const wx = dx[s] + bx - dx[t];
          const wy = dy[s] + by - dy[t];
          if (wx !== 0) wrapH = true;
          if (wy !== 0) wrapV = true;
        } else if (rank[rs] < rank[rt]) {
          parent[rs] = rt;
          dx[rs] = dx[t] - bx - dx[s];
          dy[rs] = dy[t] - by - dy[s];
        } else {
          parent[rt] = rs;
          dx[rt] = dx[s] + bx - dx[t];
          dy[rt] = dy[s] + by - dy[t];
          if (rank[rs] === rank[rt]) rank[rs] += 1;
        }
      }
      if (wrapH || wrapV) return (k + 1) / n;
    }
    return 1.0;
  }

  function meanSe(ps) {
    const m = ps.length;
    const mean = ps.reduce((a, b) => a + b, 0) / m;
    let varSum = 0;
    for (const p of ps) varSum += (p - mean) ** 2;
    const sigma = Math.sqrt(varSum / (m - 1));
    return { mean, se: sigma / Math.sqrt(m), sigma };
  }

  function wlsThroughOrigin(x, y, se) {
    const w = se.map((s) => (s > 0 ? 1 / (s * s) : 0));
    let den = 0;
    let num = 0;
    for (let i = 0; i < x.length; i++) {
      den += w[i] * x[i] * x[i];
      num += w[i] * x[i] * y[i];
    }
    if (den <= 0) return { c: 0, seC: Infinity, chi2: Infinity, r2: 0 };
    const c = num / den;
    const seC = Math.sqrt(1 / den);
    let chi2 = 0;
    let ssRes = 0;
    let ySum = 0;
    for (const yi of y) ySum += yi;
    const yMean = ySum / y.length;
    let ssTot = 0;
    for (let i = 0; i < x.length; i++) {
      const ri = y[i] - c * x[i];
      chi2 += w[i] * ri * ri;
      ssRes += ri * ri;
      const d = y[i] - yMean;
      ssTot += d * d;
    }
    return { c, seC, chi2, r2: ssTot ? 1 - ssRes / ssTot : 0 };
  }

  function seNuFromChi2(chi2Of, nuHat, chi2Min) {
    const target = chi2Min + 1;
    let left = chi2Of[0][0];
    let right = chi2Of[chi2Of.length - 1][0];
    let prevNu = chi2Of[0][0];
    let prevChi = chi2Of[0][1];
    let foundLeft = false;
    let foundRight = false;
    for (let i = 1; i < chi2Of.length; i++) {
      const nu = chi2Of[i][0];
      const chi2 = chi2Of[i][1];
      if (!foundLeft && prevNu <= nuHat && prevChi >= target && target > chi2) {
        const denom = prevChi - chi2;
        const frac = denom ? (prevChi - target) / denom : 0;
        left = prevNu + frac * (nu - prevNu);
        foundLeft = true;
      }
      if (!foundRight && prevNu >= nuHat && prevChi < target && target <= chi2) {
        const denom = chi2 - prevChi;
        const frac = denom ? (target - prevChi) / denom : 0;
        right = prevNu + frac * (nu - prevNu);
        foundRight = true;
      }
      prevNu = nu;
      prevChi = chi2;
    }
    return Math.max(nuHat - left, right - nuHat);
  }

  function fitA(sizes, pcs, ses) {
    const y = pcs.map((pc) => pc - PC_INF);
    const chi2Of = [];
    let best = null;
    for (let i = 0; i <= (2.5 - 0.5) / 0.002; i++) {
      const nu = 0.5 + i * 0.002;
      const x = sizes.map((L) => L ** (-1 / nu));
      const fit = wlsThroughOrigin(x, y, ses);
      chi2Of.push([nu, fit.chi2]);
      if (!best || fit.chi2 < best.chi2) {
        best = { chi2: fit.chi2, nu, c: fit.c, seC: fit.seC, r2: fit.r2 };
      }
    }
    const seNu = seNuFromChi2(chi2Of, best.nu, best.chi2);
    return {
      nu: best.nu,
      seNu,
      c: best.c,
      seC: best.seC,
      chi2: best.chi2,
      r2: best.r2,
      relErr: Math.abs(best.nu - NU_THEORY) / NU_THEORY,
    };
  }

  function sameSignDeltas(pcs) {
    const deltas = pcs.map((pc) => pc - PC_INF);
    if (deltas.some((d) => d === 0)) return false;
    const signs = deltas.map((d) => d > 0);
    return signs.every(Boolean) || signs.every((s) => !s);
  }

  function approachesPc(sizes, pcs, ses) {
    const absD = pcs.map((pc) => Math.abs(pc - PC_INF));
    for (let i = 0; i < sizes.length - 1; i++) {
      const slack = 2 * (ses[i] + ses[i + 1]);
      if (absD[i + 1] > absD[i] + slack) return false;
    }
    return true;
  }

  function classify(fit, pcs, ses) {
    const powered = fit.seNu <= NU_TOLERANCE * NU_THEORY;
    const r2Ok = fit.r2 >= R2_MIN;
    const slopeSig = Math.abs(fit.c) > 2 * fit.seC;
    const within = fit.relErr <= NU_TOLERANCE;
    const signOk = sameSignDeltas(pcs);
    const mono = pcs.length === FIT_SIZES.length ? approachesPc(FIT_SIZES, pcs, ses) : signOk;
    if (!signOk || !powered || !slopeSig || !r2Ok) return "INCONCLUSIVE";
    if (within && mono) return "CONFIRMED";
    if (!within && Math.abs(fit.nu - NU_THEORY) > 2 * fit.seNu) return "FALSIFIED";
    return "INCONCLUSIVE";
  }

  async function runPercolationFss(emit) {
    emit({
      type: "line",
      text: "Crosscheck: p-b-habitat-percolation-ecology-fss",
    });
    emit({ type: "line", text: ESTIMATOR_SENTENCE });
    emit({
      type: "line",
      text: `Estimator id: ${ESTIMATOR_ID} (do not switch to horizontal-only wrap without refitting c)`,
    });
    emit({
      type: "line",
      text: `Theory: p_c(inf)=${PC_INF}, nu=${NU_THEORY.toFixed(4)}. Gate: 15% on nu, R² ≥ ${R2_MIN}.`,
    });
    emit({
      type: "line",
      text: `Params: L in [${FIT_SIZES.join(", ")}], samples/L=${N_SAMPLES}, seed=${SEED}. Result is the fit, not a preset.`,
    });
    emit({ type: "line", text: "" });

    const rng = mulberry32(SEED);
    const rows = [];
    const total = FIT_SIZES.length * N_SAMPLES;
    let done = 0;
    for (let i = 0; i < FIT_SIZES.length; i++) {
      const L = FIT_SIZES[i];
      const ps = [];
      for (let t = 0; t < N_SAMPLES; t++) {
        ps.push(firstWrapEither(L, rng));
        done += 1;
        if (t % 2 === 1) {
          emit({ type: "progress", pct: Math.round((100 * done) / total) });
          await yieldToBrowser();
        }
      }
      const { mean, se, sigma } = meanSe(ps);
      const delta = mean - PC_INF;
      const sign = delta >= 0 ? "+" : "";
      emit({
        type: "line",
        text:
          `  L=${String(L).padStart(4)}  p_c_hat=${mean.toFixed(5)} ± ${se.toFixed(5)}  ` +
          `delta=${sign}${delta.toFixed(5)}  sigma=${sigma.toFixed(5)}`,
      });
      rows.push({ L, mean, se, sigma });
    }

    const pcs = rows.map((row) => row.mean);
    const ses = rows.map((row) => row.se);
    const fit = fitA(FIT_SIZES, pcs, ses);
    const result = classify(fit, pcs, ses);

    emit({ type: "line", text: "" });
    emit({
      type: "line",
      text:
        `Fit A  nu = ${fit.nu.toFixed(4)} ± ${fit.seNu.toFixed(4)}   ` +
        `c = ${fit.c >= 0 ? "+" : ""}${fit.c.toFixed(4)} ± ${fit.seC.toFixed(4)}`,
    });
    emit({
      type: "line",
      text:
        `  R² = ${fit.r2.toFixed(4)}   rel err vs 4/3 = ${(100 * fit.relErr).toFixed(1)}%  (tolerance 15%)`,
    });
    if (result === "INCONCLUSIVE") {
      emit({
        type: "line",
        text: "INCONCLUSIVE means this run is underpowered or the fit is unstable. It is not a claim that percolation has the wrong exponent.",
      });
    } else if (result === "FALSIFIED") {
      emit({
        type: "line",
        text: "FALSIFIED means this run's fit disagrees with nu=4/3 past the gate. That is a protocol failure, not a disproof of percolation.",
      });
    } else {
      emit({
        type: "line",
        text: "CONFIRMED means this run's weighted fit recovered nu within 15% of 4/3, with a tight enough error bar.",
      });
    }
    emit({ type: "line", text: `RESULT: ${result}` });
    emit({ type: "progress", pct: 100 });
    emit({ type: "result", result });
    return { passed: result === "CONFIRMED", result, fit, rows };
  }

  const root = typeof window !== "undefined" ? window : globalThis;
  root.CrosscheckRuns = root.CrosscheckRuns || {};
  root.CrosscheckRuns["p-b-habitat-percolation-ecology-fss"] = runPercolationFss;
})();
